libs/vkd3d/device.c

```diff
@@ -4987,6 +4995,72 @@ struct d3d12_device *unsafe_impl_from_ID3D12Device(d3d12_device_iface *iface)
     return impl_from_ID3D12Device(iface);
 }
 
+static inline struct d3d12_device *d3d12_device_from_ID3D12DeviceExt(ID3D12DeviceExt *iface)
+{
+    struct d3d12_device_vkd3d_ext *d12device_ext = CONTAINING_RECORD(iface, struct d3d12_device_vkd3d_ext, ID3D12DeviceExt_iface);
+    return CONTAINING_RECORD(d12device_ext, struct d3d12_device, device_vkd3d_ext);
+}
+
+static ULONG STDMETHODCALLTYPE d3d12_device_vkd3d_ext_AddRef(ID3D12DeviceExt *iface)
+{
+    struct d3d12_device *device = d3d12_device_from_ID3D12DeviceExt(iface);
+    return d3d12_device_add_ref(device);
+}
+
+static ULONG STDMETHODCALLTYPE d3d12_device_vkd3d_ext_Release(ID3D12DeviceExt *iface)
+{
+    struct d3d12_device *device = d3d12_device_from_ID3D12DeviceExt(iface);
+    return d3d12_device_release(device);
+}
+
+static HRESULT STDMETHODCALLTYPE d3d12_device_vkd3d_ext_QueryInterface(ID3D12DeviceExt *iface,
+        REFIID iid, void **out)
+{
+    struct d3d12_device *device = d3d12_device_from_ID3D12DeviceExt(iface);
+    TRACE("iface %p, iid %s, out %p.\n", iface, debugstr_guid(iid), out);
+
+    if (IsEqualGUID(iid, &IID_ID3D12DeviceExt) ||
+        IsEqualGUID(iid, &IID_IUnknown))
+    {
+        d3d12_device_vkd3d_ext_AddRef(iface);
```

Comment:
This violates the COM rule of:

> For any given object instance, a call to QueryInterface with IID_IUnknown must always return the same physical pointer value.

Wouldn't it work to just chain directly to d3d12_device_QueryInterface?

Reply:
Good catch.!